server-mux.go

```diff
@@ -0,0 +1,151 @@
+/*
+ * Minio Cloud Storage, (C) 2016 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"crypto/tls"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+var defaultHTTPMethods = []string{
+	"OPTIONS",
+	"GET",
+	"HEAD",
+	"POST",
+	"PUT",
+	"DELETE",
+	"TRACE",
+	"CONNECT",
+}
+
+// MuxConn
+type MuxConn struct {
+	net.Conn
+	buffer  []byte
+	reset   bool
+	pointer int
+	peekLen int
+}
+
+func NewMuxConn(c net.Conn) *MuxConn {
+	maxLen := 0
+	for _, m := range defaultHTTPMethods {
+		if maxLen < len(m) {
+			maxLen = len(m) + 1
+		}
+	}
+	return &MuxConn{Conn: c, peekLen: maxLen, buffer: make([]byte, maxLen)}
+}
+
+func (c *MuxConn) peekProtocol() (string, error) {
+	n, err := c.Conn.Read(c.buffer)
+	if n == 0 || err != nil {
+		return "", err
+	}
+	c.reset = true
+	for _, m := range defaultHTTPMethods {
+		if strings.HasPrefix(string(c.buffer), m+" ") {
```

Comment:
> Don't understand the logic here.. you might have to write a proper comment talking about the HTTP headers themselves.

Ah the TLS part i see how it is doing, you will be able to peek only HTTP not HTTPS.